core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1452,12 +1520,34 @@ class Log(@volatile private var _dir: File,
           latestTimestampAndOffset.offset,
           epochOptional))
       } else {
-        // Cache to avoid race conditions. `toBuffer` is faster than most alternatives and provides
-        // constant time access while being safe to use with concurrent collections unlike `toArray`.
-        val segmentsCopy = logSegments.toBuffer
-        // We need to search the first segment whose largest timestamp is >= the target timestamp if there is one.
-        val targetSeg = segmentsCopy.find(_.largestTimestamp >= targetTimestamp)
-        targetSeg.flatMap(_.findOffsetByTimestamp(targetTimestamp, logStartOffset))
+        var isFirstSegment = false
+        val targetSeg: Option[LogSegment] = {
+          // Get all the segments whose largest timestamp is smaller than target timestamp
+          val earlierSegs = segmentsCopy.takeWhile(_.largestTimestamp < targetTimestamp)
+          // We need to search the first segment whose largest timestamp is greater than the target timestamp if there is one.
+          if (earlierSegs.length < segmentsCopy.length) {
+            isFirstSegment = earlierSegs.isEmpty
+            Some(segmentsCopy(earlierSegs.length))
+          } else {
+            None
+          }
+        }
+
+        if (isFirstSegment && remoteLogManager.isDefined) {
```

Comment:
I find this condition and the logic inside this block a bit confusing.
Based on the logic below, it seems either the localOffset or the remoteOffset can be the larger value.
Under what circumstances would we have a different localOffset and remoteOffset?
Also, does this block also apply when isFirstSegment == false?

Reply:
According to JavaDocs, the API should return the offset
> of the first message whose timestamp is greater than or equals to the target timestamp

Let us say, a ListOffsetsRequest comes in for the timestamp `1450`. We will go through the local log segments in order of their start offsets, and identify the first segment that has its highest timestamp >= 1450 *(we will refer to this as our selection criteria henceforth)*. There are some possibilities here -

## Scenario 1
### Remote Segments
```
1000-1400
1400-1410
1411-1440
```

### Segments on disk
```
1411-1440
1441-1456
1457-1480
```

In this scenario, the first log segment on disk does not fulfill our *selection criteria*, the second one does. Hence, we know for sure that there are no remote logs that can possibly meet the *selection criteria* which are not also present on the disk. We definitely know that the segment `1441-1456` contains our answer, and we rely solely on it. We don't need to check the remote tier because we know that there are no segments there that are not present on disk, and are possible candidates to have an answer.

## Scenario 2
### Remote Segments
```
1000-1400
1400-1410
1411-1440
1441-1456
```

### Segments on disk
```
1457-1480
1481-1500
```

In this scenario, the first log on disk `1457-1480` fulfills our *selection criteria*. So it is possible (but not guaranteed) for the remote tier to have some segments that have been deleted from disk that also match our *selection criteria* (in this case, `1441-1456`). In this scenario, we need to check if such a segment exists in the remote tier, and if it does, we need to check its offset.

To summarize: If the very first log segment on disk matches the *selection criteria*, then it is possible that either remote or local disk has the desired offset. If not, the local disk definitely has it. In the first scenario, it is possible to find answers both in the remote and local tier, and we should select whichever is smaller.

I too do not understand the scenario for which remoteOffset will be > localOffset when both exist. I can't think of such a scenario (maybe during unclean leader election?). Regardless, the logic sounds sound, and should just return the offset that was smaller.

We can probably check if there is scope for simplification of logic at a later point in our development. I think there is definitely some code here that can benefit from refactoring / simplification.

Reply:
We also need to test the scenario as to what happens for a ListOffsetRequests call when an unclean leader election causes epoch divergence, but RLM hasn't yet deleted the remote log segments. The call to `RLM::findOffsetByTimestamp` does not use an epoch, so is it possible that it will return an incorrect offset that has been discarded due to log divergence?

Reply:
Discussed offline. Will leave a TODO comment here with a note to validate the unclean leader election scenario during testing, or add an integration test.